### PR TITLE
Updated SDK with allowed IP instance policy

### PR DESCRIPTION
**Pull request by gspurthi**
Associated with the issue: 154

Changes included in the PR: Adding support to create and get allowedIP instance policy for go sdk.

### CODE REVIEW

File: instances.go [hunk 1]
```diff
@@ -91,6 +98,24 @@ func (c *Client) GetAllowedNetworkInstancePolicy(ctx context.Context) (*Instance
 	return &policyResponse.Policies[0], nil
 }
 
+// GetAllowedIPInstancePolicy retrieves the allowed IP instance policy details associated with the instance.
+// For more information can refer the Key Protect docs in the link below:
+//<KP Link>
```
Comment (pbavari) [inline instances.go:103]: Link needs to be updated.

File: instances.go [hunk 2]
```diff
@@ -170,6 +195,36 @@ func (c *Client) SetDualAuthInstancePolicy(ctx context.Context, enable bool) err
 	return err
 }
 
+// SetAllowedIPInstancePolices updates the allowed IP instance policy details associated with an instance.
+// For more information can refet to the Key Protect docs in the link below:
+// <KP link>
```
Comment (pbavari) [inline instances.go:200]: Missing link

File: instances.go [hunk 3]
```diff
@@ -170,6 +195,36 @@ func (c *Client) SetDualAuthInstancePolicy(ctx context.Context, enable bool) err
...
+func (c *Client) SetAllowedIPInstancePolicy(ctx context.Context, enable bool, allowedIPs []string) error {
+	policy := InstancePolicy{
+		PolicyType: AllowedIP,
+		PolicyData: PolicyData{
+			Enabled:    &enable,
+			Attributes: &Attributes{},
+		},
+	}
+	if len(allowedIPs) != 0 {
+		policy.PolicyData.Attributes.AllowedIP = allowedIPs
```
Comment (pbavari) [inline instances.go:210]: validate the format of the IPs

Comment (stephaniegalang-ibm) [reply instances.go:210]: Some validation from the API side are as follows:
- attributes stuct must be omitted if `enable` == `false`. This implies that you should error if `enable` == `false` and `len(allowedIPs)` != 0
- len(allowedIPs) must be greater than 0 if `enable` == `true`
- all allowedIP items must be in ipv4 or ipv6 networks specified in CIDR format (X.X.X.X/N, X.X.X.X.X.X.X.X/N)

Comment (gspurthi) [reply instances.go:210]: Stephanie, I set the condition to verify enable and Length of allowedIPs. I believe the CIDR format of the ips is verified by key protect service. So I would not rather add that check in SDK. I can add it in CLI.

Comment (gspurthi) [reply instances.go:210]: I will be updating the document link for allowed ip policy, I see that the required
format for IPs has been mentioned there.

File: instances.go [hunk 4]
```diff
@@ -170,6 +195,36 @@ func (c *Client) SetDualAuthInstancePolicy(ctx context.Context, enable bool) err
...
+	policy := InstancePolicy{
+		PolicyType: AllowedIP,
+		PolicyData: PolicyData{
+			Enabled: &enable,
+		},
+	}
+	if enable && len(allowedIPs) > 0 {
+		policy.PolicyData.Attributes = &Attributes{}
+		policy.PolicyData.Attributes.AllowedIP = allowedIPs
+	}
```
Comment (mrodden) [inline instances.go:211]: Lets add an `else` block return an Error in the case where the user is specifying `enable=True` and `len(allowedIps) == 0` which will always be an error case.

Also we should record in a comment that we are relying on the server to validate IP address formats and other validations.

**Comment by gspurthi** [comment]
@pbavari  for the endpoint that fetched the port number, I will create a separate PR.
Please review and approve this one.

**Review by pbavari — APPROVED** [review]
Approving but this should not be published yet - need to add support for get allowed ip port api.

**Comment by gspurthi** [comment]
allowed IP port api PR for SDK: https://github.com/IBM/keyprotect-go-client/pull/42

File: kp_test.go [hunk 5]
```diff
@@ -1626,6 +1626,49 @@ func TestSetAndGetInstancePolicies(t *testing.T) {
...
+		"resources": [{
+			"creationDate": "2020-09-01T17:28:28Z",
+			"createdBy": "IBMid-50BE1MTM26",
+			"updatedBy": "IBMid-50BE1MTM26",
+			"lastUpdated": "2020-09-02T17:20:21Z",
+			"policy_type": "allowedIP",
+			"policy_data": {
+				"enabled": true,
+				"attributes": {
+					"allowed_ip": ["47.220.195.239/32", "12.184.193.84/30"]
```
Comment (stephaniegalang-ibm) [inline kp_test.go:1644]: We should avoid referencing real ip addresses. I've provided some documentation ones
```suggestion
					"allowed_ip": ["198.51.100.0/32", "2001:db8::/128"]
```

File: instances.go [hunk 6]
```diff
@@ -170,6 +196,39 @@ func (c *Client) SetDualAuthInstancePolicy(ctx context.Context, enable bool) err
...
+			Enabled: &enable,
+		},
+	}
+
+	// The IP address validation is performed by the key protect service.
+	if enable && len(allowedIPs) != 0 {
+		policy.PolicyData.Attributes = &Attributes{}
+		policy.PolicyData.Attributes.AllowedIP = allowedIPs
+	} else if enable && len(allowedIPs) == 0 {
+		return fmt.Errorf("Please provide atleast 1 IP Address")
```
Comment (stephaniegalang-ibm) [inline instances.go:215]:
```suggestion
		return fmt.Errorf("Please provide at least 1 IP subnet specified with CIDR notation")
```

File: instances.go [hunk 7]
```diff
@@ -170,6 +196,39 @@ func (c *Client) SetDualAuthInstancePolicy(ctx context.Context, enable bool) err
...
+	} else if !enable && len(allowedIPs) != 0 {
+		return fmt.Errorf("Provide IP Addresses only of the policy is being enabled")
```
Comment (stephaniegalang-ibm) [inline instances.go:217]:
```suggestion
		return fmt.Errorf("Allowed IP array should only be provided if the policy is being enabled")
```

File: kp_test.go [hunk 8]
```diff
@@ -1626,6 +1626,49 @@ func TestSetAndGetInstancePolicies(t *testing.T) {
...
+			}
+		}]
+	}`)
+
+	gock.New("http://example/com").
+		Put("/api/v2/instance/policies").
+		MatchParam("policy", "allowedIP").
+		Reply(204)
+
+	err = c.SetAllowedIPInstancePolicy(context.Background(), true, []string{"21.54.123.68/30", "53.114.213.84/32"})
```
Comment (stephaniegalang-ibm) [inline kp_test.go:1655]: Again, should remove references to real IPs

```suggestion
	err = c.SetAllowedIPInstancePolicy(context.Background(), true, []string{"192.0.2.0/24", 203.0.113.0/32"})
```

**Review by stephaniegalang-ibm — CHANGES_REQUESTED** [review]
Some suggested message wording, particularly regarding how the ip array is of ip subnets, not ip addresses

File: instances.go [hunk 9]
```diff
@@ -170,6 +196,39 @@ func (c *Client) SetDualAuthInstancePolicy(ctx context.Context, enable bool) err
...
+			Enabled: &enable,
+		},
+	}
+
+	// The IP address validation is performed by the key protect service.
+	if enable && len(allowedIPs) != 0 {
+		policy.PolicyData.Attributes = &Attributes{}
+		policy.PolicyData.Attributes.AllowedIP = allowedIPs
+	} else if enable && len(allowedIPs) == 0 {
+		return fmt.Errorf("Please provide atleast 1 IP subnet specified with CIDR notation")
```
Comment (mrodden) [inline instances.go:215]:
```suggestion
		return fmt.Errorf("Please provide at least 1 IP subnet specified with CIDR notation")
```

File: kp_test.go [hunk 10]
```diff
@@ -1626,6 +1626,49 @@ func TestSetAndGetInstancePolicies(t *testing.T) {
...
+	assert.NoError(t, err)
+
+	gock.New("http://example.com").
+		Get("/api/v2/instance/policies").
+		MatchParam("policy", "allowedIP").
+		Reply(200).
+		Body(bytes.NewReader(allowedIPPolicy))
+
+	ip, err := c.GetAllowedIPInstancePolicy(context.Background())
+
```
Comment (mrodden) [inline kp_test.go:1666]: Lets add testcases for the error paths above

1) enable=True, zero ips in list -> should return error
2) enable=False, have ips in list -> should return error

File: kp_test.go [hunk 11]
```diff
@@ -1626,9 +1626,68 @@ func TestSetAndGetInstancePolicies(t *testing.T) {
...
 	assert.True(t, gock.IsDone(), "Expected HTTP requests not called!")
 }
 
+// TestSetAllowedIPPolicyError tests the error scenarios while setting Allowed IP policy
+func TestSetAllowedIPPolicyError(t *testing.T) {
+	c, _, err := NewTestClient(t, nil)
+	c.tokenSource = &FakeTokenSource{}
+
+	err = c.SetAllowedIPInstancePolicy(context.Background(), false, []string{"192.0.2.0/24", "203.0.113.0/32"})
```
Comment (mrodden) [inline kp_test.go:1680]: Cool

**Review by mrodden — APPROVED** [review]

File: kp_test.go [hunk 12]
```diff
@@ -1626,9 +1626,68 @@ func TestSetAndGetInstancePolicies(t *testing.T) {
...
+
+	assert.Error(t, err)
+	assert.Equal(t, err.Error(), "IP address list should only be provided if the policy is being enabled")
```
Comment (mrodden) [inline kp_test.go:1683]:
```suggestion
	assert.Equal(t, "IP address list should only be provided if the policy is being enabled", err.Error())
```

Expected value comes first

File: kp_test.go [hunk 13]
```diff
@@ -1626,9 +1626,68 @@ func TestSetAndGetInstancePolicies(t *testing.T) {
...
+
+	err = c.SetAllowedIPInstancePolicy(context.Background(), true, []string{})
+
+	assert.Error(t, err)
+	assert.Equal(t, err.Error(), "Please provide atleast 1 IP subnet specified with CIDR notation")
```
Comment (mrodden) [inline kp_test.go:1688]:
```suggestion
	assert.Equal(t, "Please provide at least 1 IP subnet specified with CIDR notation", err.Error())
```

**Review by stephaniegalang-ibm — APPROVED** [review]

**Comment by mrodden** [comment]
I'm going to squash and merge since we have a lot of small fixups in the chain here.

**Pull request merged**